### PR TITLE
Prevent duplicates in Batches and Sites

### PR DESCRIPTION
**Pull request by bes-alielzein**
** This PR can be reviewed commit by commit **

The issue with the duplicated keys in the FE was actually a data issue. When seeding the database, the `tree_type` for the generated batches where randomized without making sure a given tree_type was not already used for a given batch. This resulted in duplicated `tree_type` in a batch.

For example (Black Maple is duplicated here):
image

In `batchspecies` table (`batch_id`: 10 & `tree_type_id`: 2 are duplicated across two entries):
image

To fix this issue:
- The seeding script was modified to use unique `tree_type`, `fertilizer_type` and `mulch_layer_type` per batches and sites.
- Unique constraints were added to tables where duplication could happen.

### CODE REVIEW

File: canopeum_backend/canopeum_backend/models.py [hunk 1]
```diff
@@ -121,6 +121,13 @@ class Batchfertilizer(models.Model):
     batch = models.ForeignKey(Batch, models.CASCADE, blank=True, null=True)
     fertilizer_type = models.ForeignKey(Fertilizertype, models.DO_NOTHING, blank=True, null=True)
 
+    class Meta:
+        constraints: ClassVar[list[models.UniqueConstraint]] = [
+            models.UniqueConstraint(
+                fields=["batch", "fertilizer_type"], name="unique_fertilizer_per_batch"
+            ),
+        ]
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:129]:
```suggestion
        constraints = (
            models.UniqueConstraint(
                fields=["batch", "fertilizer_type"], name="unique_fertilizer_per_batch"
            ),
        )
```

File: canopeum_backend/canopeum_backend/models.py [hunk 2]
```diff
@@ -137,6 +144,13 @@ class Batchmulchlayer(models.Model):
     batch = models.ForeignKey(Batch, models.CASCADE, blank=True, null=True)
     mulch_layer_type = models.ForeignKey(Mulchlayertype, models.DO_NOTHING, blank=True, null=True)
 
+    class Meta:
+        constraints: ClassVar[list[models.UniqueConstraint]] = [
+            models.UniqueConstraint(
+                fields=["batch", "mulch_layer_type"], name="unique_mulch_layer_per_batch"
+            ),
+        ]
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:152]:
```suggestion
        constraints = (
            models.UniqueConstraint(
                fields=["batch", "mulch_layer_type"], name="unique_mulch_layer_per_batch"
            ),
        )
```

File: canopeum_backend/canopeum_backend/models.py [hunk 3]
```diff
@@ -154,17 +168,34 @@ class BatchSpecies(models.Model):
     tree_type = models.ForeignKey(Treetype, models.DO_NOTHING, blank=True, null=True)
     quantity = models.IntegerField(blank=True, null=True)
 
+    class Meta:
+        constraints: ClassVar[list[models.UniqueConstraint]] = [
+            models.UniqueConstraint(fields=["batch", "tree_type"], name="unique_species_per_batch"),
+        ]
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:174]:
```suggestion
        constraints = (
            models.UniqueConstraint(fields=["batch", "tree_type"], name="unique_species_per_batch"),
        )
```

File: canopeum_backend/canopeum_backend/models.py [hunk 4]
```diff
@@ -154,17 +168,34 @@ class BatchSpecies(models.Model):
...
 
 class BatchSeed(models.Model):
     batch = models.ForeignKey(Batch, models.CASCADE, blank=True, null=True)
     tree_type = models.ForeignKey(Treetype, models.DO_NOTHING, blank=True, null=True)
     quantity = models.IntegerField(blank=True, null=True)
 
+    class Meta:
+        constraints: ClassVar[list[models.UniqueConstraint]] = [
+            models.UniqueConstraint(fields=["batch", "tree_type"], name="unique_seed_per_batch"),
+        ]
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:185]:
```suggestion
        constraints = (
            models.UniqueConstraint(fields=["batch", "tree_type"], name="unique_seed_per_batch"),
        )
```

File: canopeum_backend/canopeum_backend/models.py [hunk 5]
```diff
@@ -154,17 +168,34 @@ class BatchSpecies(models.Model):
...
 class BatchSupportedSpecies(models.Model):
     batch = models.ForeignKey(Batch, models.CASCADE, blank=True, null=True)
     tree_type = models.ForeignKey(Treetype, models.DO_NOTHING, blank=True, null=True)
 
+    class Meta:
+        constraints: ClassVar[list[models.UniqueConstraint]] = [
+            models.UniqueConstraint(
+                fields=["batch", "tree_type"], name="unique_supported_species_per_batch"
+            ),
+        ]
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:197]:
```suggestion
        constraints = (
            models.UniqueConstraint(
                fields=["batch", "tree_type"], name="unique_supported_species_per_batch"
            ),
        )
```

File: canopeum_backend/canopeum_backend/models.py [hunk 6]
```diff
@@ -332,6 +363,13 @@ class Sitetreespecies(models.Model):
     tree_type = models.ForeignKey("Treetype", models.DO_NOTHING, blank=True, null=True)
     quantity = models.IntegerField(blank=True, null=True)
 
+    class Meta:
+        constraints: ClassVar[list[models.UniqueConstraint]] = [
+            models.UniqueConstraint(
+                fields=["site", "tree_type"], name="unique_tree_species_per_site"
+            ),
+        ]
```
Comment (Samuel-Therrien-Beslogic) [inline canopeum_backend/canopeum_backend/models.py:371]:
```suggestion
        constraints = (
            models.UniqueConstraint(
                fields=["site", "tree_type"], name="unique_tree_species_per_site"
            ),
        )
```

**Review by Samuel-Therrien-Beslogic — CHANGES_REQUESTED** [review]
If you use immutable types, you won't need to explicitely declare those as `ClassVar`s.
https://docs.astral.sh/ruff/rules/mutable-class-default/
> When mutable values are intended, they should be annotated with typing.ClassVar. When mutability is not required, values should be immutable types, like tuple or frozenset.

File: canopeum_backend/canopeum_backend/models.py (hunk 1 above)
Comment (bes-alielzein) [reply canopeum_backend/canopeum_backend/models.py]: Fixed

File: canopeum_backend/canopeum_backend/models.py (hunk 2 above)
Comment (bes-alielzein) [reply canopeum_backend/canopeum_backend/models.py]: Fixed

File: canopeum_backend/canopeum_backend/models.py (hunk 4 above)
Comment (bes-alielzein) [reply canopeum_backend/canopeum_backend/models.py]: Fixed

File: canopeum_backend/canopeum_backend/models.py (hunk 6 above)
Comment (bes-alielzein) [reply canopeum_backend/canopeum_backend/models.py]: Fixed

File: canopeum_backend/canopeum_backend/models.py (hunk 5 above)
Comment (bes-alielzein) [reply canopeum_backend/canopeum_backend/models.py]: Fixed

File: canopeum_backend/canopeum_backend/models.py (hunk 3 above)
Comment (bes-alielzein) [reply canopeum_backend/canopeum_backend/models.py]: Fixed

**Review by Samuel-Therrien-Beslogic — APPROVED** [review]

**Pull request merged**